### ISSUE TITLE
[BUG] incorrect CSEQ in ACK (case - processing 200ok retransmits on local generated reINVITE)

### ISSUE THREAD

**Issue by Oleg-voip** [OpenSIPS/opensips#2106]
**OpenSIPS version you are running**

```
version: opensips 3.0.2 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-FUTEX-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 20ffc6c8c
main.c compiled on  with gcc 7
```

**Describe the bug**

I use these settings:
```
modparam("dialog", "reinvite_ping_interval", 30)

and

create_dialog("BERrCc");
```
When opensips generate reINVITE by timeout 30 seconds, opensips incorrect process 200ok retransmittions. Attached pcap:
[problem-example-200-retransmittions.zip](https://github.com/OpenSIPS/opensips/files/4598643/problem-example-200-retransmittions.zip)
Check frames 34-38 and you can see, that when opensips receive 200ok retransmittion, he send ACK with increased CSEQ.

Frame 32: opensips receive 200ok.
Frame 34: opensips generate ACK with CSEQ "2 ACK"
Frame 35: callee send the retransmission of 200ok (frame 32) to opensips
Frame 36: opensips generate ACK with CSEQ "3 ACK"
Frame 37: callee send the retransmission of 200ok (frame 32) to opensips
Frame 38: opensips generate ACK with CSEQ "4 ACK"

It is incorrect.

**OpenSIPS config**
```
####### Global Parameters ##############################################
log_level=3
log_facility=LOG_LOCAL0
udp_workers=10

auto_aliases=no
mhomed=no

listen = udp:2.2.2.2:5060

server_header="Server: TEST"
user_agent_header="User-Agent: TEST"
server_signature=no

pv_print_buf_size=512000

disable_core_dump=yes
disable_503_translation=yes


####### Modules Section ################################################
mpath="/usr/lib/x86_64-linux-gnu/opensips/modules"

########################################################################
loadmodule "avpops.so"
########################################################################

########################################################################
loadmodule "cfgutils.so"
########################################################################
modparam("cfgutils", "varset", "subscriber_id=s:TEST")

########################################################################
loadmodule "dialog.so"
########################################################################
modparam("dialog", "db_mode", 0)
modparam("dialog", "hash_size", 4096)
modparam("dialog", "log_profile_hash_size", 12)
modparam("dialog", "dlg_match_mode", 1)
modparam("dialog", "default_timeout", 3600)
modparam("dialog", "reinvite_ping_interval", 30)

########################################################################
loadmodule "gflags.so"
########################################################################


########################################################################
loadmodule "proto_udp.so"
########################################################################


########################################################################
loadmodule "rr.so"
########################################################################


########################################################################
loadmodule "signaling.so"
########################################################################


########################################################################
loadmodule "sipmsgops.so"
########################################################################


########################################################################
loadmodule "sl.so"
########################################################################

########################################################################
loadmodule "tm.so"
########################################################################
modparam("tm", "fr_inv_timeout", 240)
modparam("tm", "restart_fr_on_each_reply", 0)
modparam("tm", "onreply_avp_mode", 1)
modparam("tm", "timer_partitions", 4)
modparam("tm", "disable_6xx_block", 1)
modparam("tm", "auto_100trying", 0)
modparam("tm", "ruri_matching", 0)
modparam("tm", "via1_matching", 0)


########################################################################
loadmodule "topology_hiding.so"
########################################################################
modparam("topology_hiding", "th_callid_passwd", "TEST")
modparam("topology_hiding", "th_callid_prefix", "TEST")


########################################################################
loadmodule "uac.so"
########################################################################


route {
    if (has_totag()) {
        if (topology_hiding_match()) {
            if (!validate_dialog()) fix_route_dialog();
            if (is_method("INVITE|UPDATE")) t_on_reply("SEQ_REPLY");
            route(RELAY);
        } else {
            if (is_method("ACK|BYE")) route(CHECK_TRANS);
            send_reply(481, "Call Leg/Transaction Does Not Exist");
        }
    }

    if (is_method("CANCEL")) {
        route(CHECK_TRANS);
        exit;
    }

    if (is_method("INVITE")) route(INVITE);

    send_reply(481, "Call Leg/Transaction Does Not Exist");
}

route[INVITE] {
    if (!t_check_trans()) t_newtran();
    send_reply(100, "Trying");
    create_dialog("BERrCc");
    topology_hiding("C");
    $du = "sip:3.3.3.3:5060";

    route(RELAY);
}

onreply_route[SEQ_REPLY] {
    $var(test) = 1;
}

onreply_route[TRUNK_REPLY] {
    $var(test) = 1;
}

failure_route[TRUNK_FAILED] {
    $var(test) = 1;
}

route[CHECK_TRANS] {
    if (t_check_trans()) {
        route(RELAY);
    } else {
       if (is_method("CANCEL")) send_reply(481, "Call Leg/Transaction Does Not Exist");
    }
    return;
}

route[RELAY] {
    if (!t_relay()) {
        sl_send_reply(500, "Internal Error");
    }
    exit;
}

local_route {
    if (is_method("INVITE")) {
    }
}
```

**To Reproduce**

Try to use this config. Opensips must generate reINVITE over dialog module.
Send from callee retransmittions of 200ok to opensips and you will see that CSEQ in ACK incremented for every 200ok retransmittion.

**Expected behavior**

When opensips receive 200ok retransmittions to local generated reINVITE he must send ACK with the same CSEQ.

**Relevant System Logs**

Logs (level 4) of the attached pcap:
[opensips-logs.txt](https://github.com/OpenSIPS/opensips/files/4598688/opensips-logs.txt)

**OS/environment information**
 - Operating System: 
 - OpenSIPS installation: 
 - other relevant information:

Please check!
Thank you!

**Additional context**

**Comment by Oleg-voip** [comment]
Hello there, something new about the issue?

**Comment by digipigeon** [comment]
Hello,

I am sorry to report that I think that this is still problematic.

Firstly I can confirm that we have observed the problem before and after this patch was applied and the patch does fix the problem with CSeq on locally generated invites unless SIP Ping is enabled as well.

If a locally generated INVITE is sent at the same time as a SIP OPTIONS ping.

Focusing on the Upstream:

On the SIP trace we see the INVITE (CSeq 122) go out just before the SIP OPTIONS (123). 

First we get a 200OK for the INVITE (CSeq 122), __Opensips then sends an ACK (CSeq 123)__.

We get then see 200OK for the SIP OPTIONS (CSeq 123) as expected.

Note that if we look at the downstream branch we see things working in the same request / response order, however the packet counters are the same.

This was observed on version:
```
version: opensips 3.0.2 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 0e02e4afe
main.c compiled on  with gcc 7
```

**Comment by rvlad-patrascu** [comment]
Taking into account the CSEQ issue you described here and other race conditions we identified, we've decided to disallow enabling both RE-INVITE and OPTIONS pinging for the same endpoint. The motivation for not attempting a fix is that it doesn't make too much sense practically to try to support this in the first place. After all, if the endpoint responds to one of the requests (invite/options) it should be enough to determine its status.
